goben/prober.go

```diff
@@ -0,0 +1,323 @@
+	// Since sender is an interface net.Addr, we have to cast it down to net.IPAddr/net.UDPAddr type to get the IP
+	var senderIP net.IP
+	switch sender := sender.(type) {
+	case *net.IPAddr:
+		senderIP = sender.IP
+	case *net.UDPAddr:
+		senderIP = sender.IP
+	}
+
+	msg, err := icmp.ParseMessage(protocolICMP, pktbuf[:n])
+	if err != nil {
+		return nil, nil, err
+	}
+	return senderIP, msg, nil
+}
+
+// send ICMP message
+// We sent the icmp packets synchronously
+func (p *Prober) send(runID uint16, morePkts chan bool) {
+	seq := runID & uint16(0xff00)
+	for i := 0; i < p.config.pktsPerProbe; i++ {
+		target := p.config.target
+		if p.config.debug {
+			log.Printf("Request to=%s id=%d seq=%d", target, runID, seq)
+		}
+		if _, err := p.conn.WriteTo(p.packetToSend(runID, seq), parseIP(target)); err != nil {
+			log.Println(err.Error())
+			continue
+		}
+		if p.config.debug {
+			fmt.Println("SEND A ICMP PACKET")
+		}
+		morePkts <- true
+		seq++
+		time.Sleep(p.config.pktInterval)
+	}
+	if p.config.debug {
+		log.Printf("%s: Done sending packets, closing the tracker.", p.config.source)
+	}
+	close(morePkts)
+}
+
+// receive ICMP message and compute the measurements
+func (p *Prober) recv(runID uint16, morePkts chan bool) {
+	// keep track if the packet arrived has been received before
+	received := make(map[string]bool)
+	// a counter to make sure we read all packets arrived (including the outstanding
+	// ones after the sender has closed the connection)
+	outstandingPkts := 0
+	// the byte stream buffer
+	pktbuf := make([]byte, 1500)
+
+	for {
+		if outstandingPkts == 0 {
+			if _, ok := <-morePkts; ok {
+				outstandingPkts++
+			} else {
+				return
+			}
+		}
+		senderIP, msg, err := p.packetToRecv(pktbuf)
+		if err != nil {
+			log.Printf("Unmarshalling icmp message Error: %s\n", err.Error())
+			if neterr, ok := err.(*net.OpError); ok && neterr.Timeout() {
+				return
+			}
+		}
+		target := senderIP.String()
+		echoMsg, ok := msg.Body.(*icmp.Echo)
+		if !ok {
+			log.Println("Got wrong packet in ICMP echo reply.") // should never happen
+			continue
+		}
+
+		// get rtt
+		rtt := time.Since(bytesToTime(echoMsg.Data))
+
+		// check if this packet belong to this run
+		if !matchPacket(runID, echoMsg.ID, echoMsg.Seq) && p.config.debug {
+			log.Printf(
+				"Reply from=%s id=%d seq=%d rtt=%s Unmatched packet, probably from the last probe run.\n",
+				target, echoMsg.ID, echoMsg.Seq, rtt)
+			continue
+		}
+
+		// check if we have seen this packet before
+		pktID := fmt.Sprintf("%s_%d", target, echoMsg.Seq)
+		if received[pktID] && p.config.debug {
+			log.Printf("Duplicate reply from=%s id=%d seq=%d rtt=%s\n", target, echoMsg.ID, echoMsg.Seq, rtt)
+			continue
+		}
+
+		// record the rrt
+		if p.config.debug {
+			log.Printf("RTT: src=%s, dst=%s, rtt=%s\n", p.config.source, target, rtt)
+		}
+		entry := make([]string, 2)
+		entry[0] = target
+		entry[1] = rtt.Round(time.Millisecond).String() // round to millisecond
```

Comment:
Something seems to be going wrong here. The results I collect give very strange numbers in the range of seconds.

Reply:
10.0.0.2,15ms
10.0.0.2,500ms
10.0.0.2,515ms
10.0.0.2,2.5s
10.0.0.2,2.514s
10.0.0.2,3s
10.0.0.2,3.015s

For example, this does not seem right.